net: Fix variable buffer length configuration

rlubos opened the pull request:
This PR fixes build errors and net_pkt fragment allocation issue when CONFIG_NET_BUF_VARIABLE_DATA_SIZE is enabled.

The most problematic case was the `net_pkt_get_reserve_data()` function, used in `net_pkt_get_frag()` for example, which blindly assumed that fixed-sized buffers are used. The problem with this function is that it does not specify the fragment length to allocate, which did not work well in variable-length scenario. ~~Because of this, I propose a new Kconfig which allows to specify the default fragment length for this configuration: https://github.com/zephyrproject-rtos/zephyr/commit/db8b2959d216710ae4d42318eec4ad6dc09fb311~~. I propose to modify the APIs slightly, to include the information about the minimal fragment length to allocate.

~~This approach appealed to me as the lowest hanging fruit to make the variable-length configuration work. I'm open for discussion though, as I am aware the solution is not ideal. An alternative, perhaps better way to tackle the problem would be to modify certain `net_pkt` APIs, `net_pkt_get_frag()`, `net_pkt_get_reserve_tx_data()` etc. to allow to specify minimal fragment length required to allocate. This would allow to allocate the exact buffer size in variable-length scenario, and verify if the fixed-sized buffer is large enough to meet the requirements otherwise. This however indicates a stable API change including changes in multiple drivers/libraries. I'm not sure if we want to go down that road.~~

Fixes #50188

rlubos committed:
net: tcp: Fix build with CONFIG_NET_BUF_VARIABLE_DATA_SIZE

In case variable buffer length feature was enabled, the TCP stack did
not build due to unconditional references to CONFIG_NET_BUF_DATA_SIZE,
which is not avaialable in this configuration.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: pkt: Fix build with CONFIG_NET_BUF_VARIABLE_DATA_SIZE

In case variable buffer length feature was enabled, net_pkt code did
not build due to unconditional references to CONFIG_NET_BUF_DATA_SIZE,
which is not avaialable in this configuration.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: shell: Fix build with CONFIG_NET_BUF_VARIABLE_DATA_SIZE

The "net mem" command handler did not take the variable buffer length
configuration, and failed to build in such case.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: pkt: Introduce minimum length requirement to net_pkt_get_frag()

net_pkt_get_frag() and a few other functions did not specify the
allocated fragment length, incorrectly assuming that fixed-sized
buffers are always used.

In order to make the function work properly also with variable-sized
buffers, extend the function argument list with minimum expected
fragment length parameter. This allows to use net_buf_alloc_len()
allocator in variable buffer length configuration, as well as verify if
the fixed-sized buffer is large enough to satisfy the requirements
otherwise.

Update the existing codebase to provide the expected fragment length,
based on the context.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented on subsys/net/ip/Kconfig:
```diff
@@ -752,6 +752,16 @@ config NET_BUF_DATA_POOL_SIZE
 	  This value tell what is the size of the memory pool where each
 	  network buffer is allocated from.
 
+config NET_BUF_VARIABLE_DEFAULT_FRAG_SIZE
```
Yes, I was having something like this in mind (see the PR description), I just wasn't sure if it's acceptable to introduce such an API change to a stable API. Although if we can consider net_pkt API as internal, then I guess there's no concern. I can implement it this way, will try to propose something tomorrow.

jukkar replied:
One option is to leave `net_pkt_get_frag()` as is and introduce a new function that adds the len parameter, and then convert all the calls in zephyr tree to use the new one. If there are external users of the function, they can still use the old one.

rlubos replied:
I'm not sure if I'd like to introduce yet another function just to avoid API change... After thinking about it, the migration for out-of-tree users should be rather straightforward, as they could simply use `CONFIG_NET_BUF_DATA_SIZE` as the length parameter - they're bound to use fixed-sized buffers anyway, as it wouldn't work otherwise. 

I've pushed an update with a new proposal - the API is updated and (hopefully) all of the Zephyr codebase updated to provide the expected fragment length as the parameter. Most of the codebase should now be compatible with both fixed-sized and variable-sized buffers, with exception to some tests and `eth_sam_gmac.c` driver, where I've addded dependency to fixed-size configuration (similar to `eth_dwmac.c`) - simply because the driver was too much dependent on the fixed-sized buffers to sort it out w/o more serious refactoring (which I'd rather avoid w/o having hw to verify changes).

rlubos committed:
doc: release_notes: 3.3: Add net_pkt stable API change

Document the stable API change in the net_pkt API along with the
expected action from the existing applications.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar reviewed (commented):
Looks reasonable to me. I just noticed that we do not have any network tests that sets `CONFIG_NET_BUF_VARIABLE_DATA_SIZE`, probably because we were missing this functionality that actually works properly. I think we need to test this somehow. Simplest option is probably to add a test variant to relevant testcase.yaml files, not sure if there is a need to have dedicated test(s) that make sure we can send/receive data with this option enabled.

rlubos committed:
tests: net: Enable CONFIG_NET_BUF_VARIABLE_DATA_SIZE in some tests

CONFIG_NET_BUF_VARIABLE_DATA_SIZE option had not test coverage at all,
making in vulnerable for regressions. There is no dedicated test suite
to verify this option, but at least we can enable it in a few test
suites to make sure the configuration is functional.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
> Looks reasonable to me. I just noticed that we do not have any network tests that sets `CONFIG_NET_BUF_VARIABLE_DATA_SIZE`, probably because we were missing this functionality that actually works properly. I think we need to test this somehow. Simplest option is probably to add a test variant to relevant testcase.yaml files, not sure if there is a need to have dedicated test(s) that make sure we can send/receive data with this option enabled.

True, I think I'll just add an extra configuration to the tests in a few areas that were affected (6lo, tcp) to verify this variable-sized buffer configuration. I see there's no dedicated Ethernet L2 tests suite other than MGMT, but some tests like IPv6 create a dummy Ethernet interface so it could be verified in there as well.

`net_pkt` tests unfortunately do not seem like a good candidate as the whole testsuite seems to rely on the fixed buffer configuration (making use of `CONFIG_NET_BUF_DATA_SIZE` in test cases), so it'd need to undergo some more serious refactoring to verify variable length configuration.

jukkar commented:
> True, I think I'll just add an extra configuration to the tests in a few areas that were affected (6lo, tcp) to verify this variable-sized buffer configuration. I see there's no dedicated Ethernet L2 tests suite other than MGMT, but some tests like IPv6 create a dummy Ethernet interface so it could be verified in there as well.

Agreed, that was my thinking too. Basically we get the basic testing for "free" because of the internal Ethernet interface.

> `net_pkt` tests unfortunately do not seem like a good candidate as the whole testsuite seems to rely on the fixed buffer configuration (making use of `CONFIG_NET_BUF_DATA_SIZE` in test cases), so it'd need to undergo some more serious refactoring to verify variable length configuration.

Yes, I noticed it too, the tests were concentrating very much to the fixed len buffers. Those net_pkt tests could be enhanced too but that could be done in some other PR.
